[Backport v1.14-branch] tests: turn off watchdog for sam_e70 test kernel/critical

maksimmasalski opened the pull request:
Wrong configuration of a watchdog in v1.14-branch for sam_e70_xplained_board caused
endless reboot loop. During tests I found out that if to decrease
NUM_MILLISECONDS numeric constant from 5000ms to 4000ms, then
test will complete successfully. It seems that problem caused by
a watchdog.
So to solve that issue were 2 variants. First is to make constant
NUM_MILLISECONDS less than 5000, for example 3000 or 4000.
Second variant to disable a watchdog using new config for the board
CONFIG_WDT_DISABLE_AT_BOOT=y
I decided to use second variant and created a new config file
for the sam_e70_xplained hardware board.

Signed-off-by: Maksim Masalski <maksim.masalski@intel.com>

maksimmasalski commented:
@pabigot I did that fix for the board sam_e70_xplained. Please take a look.

maksimmasalski commented:
My way to fix #20603

ioannisg reviewed (changes requested):
Is it possible to just do what is there in _master_ branch?

pabigot reviewed (commented):
I don't think this is right, because it's specific to `sam_e70_xplained` when there are other SAM platforms that also have a watchdog.

The path I was thinking was to simply add `CONFIG_WDT_DISABLE_AT_BOOT=y` to `prj.conf`.  I don't get why it's necessary to test this with watchdog enabled.  That may need wider discussion about a policy for only enabling watchdog in tests where it's relevant.

Also I agree with @ioannisg: this should target master.  I don't think this is worth changing in the LTS release.  That, though, should be informed by some sort of policy.

ioannisg commented:
> Also I agree with @ioannisg: this should target master. I don't think this is worth changing in the LTS release. That, though, should be informed by some sort of policy.

@pabigot , I think I might have been using too complex and confusing syntax :(
I wanted to say that we should probably do what we already do in _master_ branch: 
Below is the test-case .yaml file.

```
tests:
  kernel.critical:
    platform_exclude: nsim_sem_mpu_stack_guard nsim_em_mpu_stack_guard
    filter: not CONFIG_WDT_SAM
  kernel.critical.sam:
    filter: CONFIG_WDT_SAM
    extra_configs:
     - CONFIG_WDT_DISABLE_AT_BOOT=y
  kernel.critical.nsim:
    platform_whitelist: nsim_sem_mpu_stack_guard nsim_em_mpu_stack_guard
    extra_configs:
      - CONFIG_TEST_HW_STACK_PROTECTION=n
```

I remember solving this issue myself for 2.0 release.

ioannisg commented:
@maksimmasalski @pabigot it is _OK_ to solve this also in the LTS - no objection to improving the test suite in 1.14.x as well.

Commit:
tests: disable watchdog for sam_e70 in test kernel/critical

Wrong configuration of a watchdog for sam_e70_xplained_board caused
endless reboot loop. It seems that problem caused by a watchdog.
I created a new config for the sam_e70_xplained hardware board
like in master branch.

Signed-off-by: Maksim Masalski <maksim.masalski@intel.com>

maksimmasalski commented:
@ioannisg I created new PR for that fix to the master. Because for that PR I used a parent v.1.14-branch, and it is unable to directly merge fix created in that branch with a master.

nashif commented:
> @ioannisg I created new PR #20646 for that fix to the master. Because for that PR I used a parent v.1.14-branch, and it is unable to directly merge fix created in that branch with a master.

@maksimmasalski  @ioannisg  is suggesting to fix this the same way it is fixed in master:

```
  kernel.critical.sam:
    filter: CONFIG_WDT_SAM
    extra_configs:
     - CONFIG_WDT_DISABLE_AT_BOOT=y

```

maksimmasalski commented:
> > @ioannisg I created new PR #20646 for that fix to the master. Because for that PR I used a parent v.1.14-branch, and it is unable to directly merge fix created in that branch with a master.
> 
> @maksimmasalski @ioannisg is suggesting to fix this the same way it is fixed in master:
> 
> ```
>   kernel.critical.sam:
>     filter: CONFIG_WDT_SAM
>     extra_configs:
>      - CONFIG_WDT_DISABLE_AT_BOOT=y
> ```

That why I decided to fix it in branch v1.14-branch, because here it is not fixed and test can't complete successfully. I think my PR#20646 is useless for master.

ioannisg commented:
>Ah, I understand idea. Just copy from master fix to the branch v1.14-branch. Hah, that makes sense.

Nice, and sorry for the confusion. You need to copy the filter to the main test-case, as well, so you can exclude the SAM platform.

ioannisg reviewed (changes requested):
Actually, now that I see this: why is it copied in the tests/kernel/common?

Also - could you copy the filter to the default test?     `filter: not CONFIG_WDT_SAM`

maksimmasalski commented:
> Actually, now that I see this: why is it copied in the tests/kernel/common?
> 
> Also - could you copy the filter to the default test? `filter: not CONFIG_WDT_SAM`

My fault, should be in critical.

maksimmasalski commented:
@ioannisg I added it, but it is not working.
Only working variant is a separate prj_sam_e70_xplained.conf file fixes that issue.

ioannisg reviewed (commented):
change is fine, but now i see that the tests have been renamed from .common to ,critical in master branch. So i wonder if you should do the same here. Now the added test-case name is inconsistent with the rest

Commit:
tests: disable watchdog for sam_e70 in test kernel/critical

Wrong configuration of a watchdog for sam_e70_xplained_board caused
endless reboot loop. It seems that problem caused by a watchdog.
I created a new config for the sam_e70_xplained hardware board
like in master branch.

Signed-off-by: Maksim Masalski <maksim.masalski@intel.com>

maksimmasalski commented:
> change is fine, but now i see that the tests have been renamed from .common to ,critical in master branch. So i wonder if you should do the same here. Now the added test-case name is inconsistent with the rest

I modified right now name of the test from criticial to common, made it aligned with another test names.

maksimmasalski commented:
@ioannisg 
Same issue on master has been fixed. Do I need just backport master patch to 1.14 branch?

stephanosio commented:
I have noticed that, for some SoCs (e.g. `wdt_sam.c`, `wdt_sam0.c`, `wdt_mchp_xec.c`), watchdog is activated during SoC reset or driver init, while, for others (e.g. `wdt_nrfx.c`, `wdt_mcux_wdog.c`), watchdog is activated after the driver `setup` function is called (i.e. not as part of OS initialisation).

For instance, the SAM 7 series SoC watchdog timer is activated by default after SoC reset because WDT_MR.WDDIS=0 on reset. while the nRF series SoC watchdog timer is deactivated by default unless the driver `setup` function is called.

**This means that we have two possible states for the watchdog after OS initialisation, across different SoCs when `CONFIG_WDT_DISABLE_AT_BOOT=n` (note that `wdt_nrfx.c`, for instance, does not enable watchdog when `CONFIG_WDT_DISABLE_AT_BOOT=n`).**

There could be two potentially feasible approaches to fixing this not-so-well-defined behaviour:

1. Enforce watchdog timer state based on `CONFIG_WDT_DISABLE_AT_BOOT` during driver init.
    - For SoCs that enable WDT by default: Disable WDT when `CONFIG_WDT_DISABLE_AT_BOOT=y`
    - For SoCs that disable WDT by default: Enable WDT when `CONFIG_WDT_DISABLE_AT_BOOT=n`
    - For both types of SoCs, in case of `CONFIG_WDT_DISABLE_AT_BOOT=n`, configure WDT using the default WDT configurations defined in Kconfig (if not defined, define them).

2. Make `CONFIG_WDT_DISABLE_AT_BOOT` `default y` and assume that WDT is disabled after OS initialisation by default (i.e. define this as the "standard" behaviour).
    - Reserve `CONFIG_WDT_DISABLE_AT_BOOT=n` as a special case for the SoCs that enable WDT by default after reset (do not enforce this as "enable at boot" for the SoCs that do not enable WDT after reset).
    - This special case serves as a fail-safe mechanism in case of an OS boot failure.
    - Leave the default WDT configurations at the mercy of the SoC default reset state. The rationale is that, if an SoC enables WDT on reset, other WDT configuration reset states would have been defined so as to allow a fail-safe reset.

While the first approach would be the most comprehensive; in terms of practicality, the second approach would make more sense.

In terms of this issue, if we take the first approach, we should `CONFIG_WDT_DISABLE_AT_BOOT=y` in `prj.conf`; for the second approach, there would be no need to specify `CONFIG_WDT_DISABLE_AT_BOOT=y` as it is `default y`.

maksimmasalski commented:
> I have noticed that, for some SoCs (e.g. `wdt_sam.c`, `wdt_sam0.c`, `wdt_mchp_xec.c`), watchdog is activated during SoC reset or driver init, while, for others (e.g. `wdt_nrfx.c`, `wdt_mcux_wdog.c`), watchdog is activated after the driver `setup` function is called (i.e. not as part of OS initialisation).
> 
> For instance, the SAM 7 series SoC watchdog timer is activated by default after SoC reset because WDT_MR.WDDIS=0 on reset, while the nRF series SoC watchdog timer is not activated by default unless the driver `setup` function is called.
> 
> **This means that we have two possible states for the watchdog after OS initialisation, across different SoCs when `CONFIG_WDT_DISABLE_AT_BOOT=n` (note that `wdt_nrfx.c`, for instance, does not enable watchdog when `CONFIG_WDT_DISABLE_AT_BOOT=n`).**
> 
> There could be two potentially feasible approaches to fixing this not-so-well-defined behaviour:
> 
>     1. Enforce watchdog timer state based on `CONFIG_WDT_DISABLE_AT_BOOT` during driver init.
>        
>        * For SoCs that enable WDT by default: Disable WDT when `CONFIG_WDT_DISABLE_AT_BOOT=y`
>        * For SoCs that disable WDT by default: Enable WDT when `CONFIG_WDT_DISABLE_AT_BOOT=n`
>        * For both types of SoCs, in case of `CONFIG_WDT_DISABLE_AT_BOOT=n`, configure WDT using the default WDT configurations defined in Kconfig (if not defined, define them).
> 
>     2. Make `CONFIG_WDT_DISABLE_AT_BOOT` `default y` and assume that WDT is disabled after OS initialisation by default (i.e. define this as the "standard" behaviour).
>        
>        * Reserve `CONFIG_WDT_DISABLE_AT_BOOT=n` as a special case for the SoCs that enable WDT by default after reset (do not enforce this as "enable at boot" for the SoCs that do not enable WDT after reset).
>        * This special case serves as a fail-safe mechanism in case of an OS boot failure.
>        * Leave the default WDT configurations at the mercy of the SoC default reset state. The rationale is that, if an SoC enables WDT on reset, other WDT configuration reset states would have been defined so as to allow a fail-safe reset.
> 
> 
> While the first approach would be the most comprehensive; in terms of practicality, the second approach would make more sense.
> 
> In terms of this issue, if we take the first approach, we should `CONFIG_WDT_DISABLE_AT_BOOT=y` in `prj.conf`; for the second approach, there would be no need to specify `CONFIG_WDT_DISABLE_AT_BOOT=y` as it is `default y`.

Thank you for such a detailed comment.
